### ISSUE TITLE
RecyclerView sample

### ISSUE THREAD

**Issue by cmelchior** [realm/realm-android-adapters#5]
We are seeing a number of support issue around RecyclerView and how to handler changelisteners and Realm updates in that case. See e.g https://github.com/realm/realm-java/issues/2408 and https://github.com/realm/realm-java/issues/2650

We should provide a sample using the RecyclerView that could illustrate how to make this work.

**Comment by Zhuinden** [comment]
I haven't tested this yet, but [based on previous incarnations, this sounds about right](http://stackoverflow.com/a/31188310/2413303).

``` java
public abstract class RealmRecyclerViewAdapter<T extends RealmObject, VH extends RecyclerView.ViewHolder>
    extends RecyclerView.Adapter<VH> { //put this in `io.realm`

    protected LayoutInflater inflater;
    protected OrderedRealmCollection<T> adapterData;
    protected Context context;
    private final RealmChangeListener listener;

    public RealmRecyclerViewAdapter(Context context, OrderedRealmCollection<T> data) {
        if (context == null) {
            throw new IllegalArgumentException("Context cannot be null");
        }
        this.context = context;
        this.adapterData = data;
        this.inflater = LayoutInflater.from(context);
        this.listener = new RealmChangeListener<RealmResults<T>>() {
            @Override
            public void onChange(RealmResults<T> results) {
                notifyDataSetChanged();
            }
        };

        if (data != null) {
            addListener(data);
        }
    }

    private void addListener(OrderedRealmCollection<T> data) {
        if (data instanceof RealmResults) {
            RealmResults realmResults = (RealmResults) data;
            realmResults.addChangeListener(listener);
        } else if (data instanceof RealmList) {
            RealmList realmList = (RealmList) data;
            realmList.realm.handlerController.addChangeListenerAsWeakReference(listener);
        } else {
            throw new IllegalArgumentException("RealmCollection not supported: " + data.getClass());
        }
    }

    private void removeListener(OrderedRealmCollection<T> data) {
        if (data instanceof RealmResults) {
            RealmResults realmResults = (RealmResults) data;
            realmResults.removeChangeListener(listener);
        } else if (data instanceof RealmList) {
            RealmList realmList = (RealmList) data;
            realmList.realm.handlerController.removeWeakChangeListener(listener);
        } else {
            throw new IllegalArgumentException("RealmCollection not supported: " + data.getClass());
        }
    }

    /**
     * Returns how many items are in the data set.
     *
     * @return the number of items.
     */
    @Override
    public int getItemCount() {
        if (adapterData == null) {
            return 0;
        }
        return adapterData.size();
    }

    /**
     * Get the data item associated with the specified position in the data set.
     *
     * @param position Position of the item whose data we want within the adapter's
     * data set.
     * @return The data at the specified position.
     */
    @Override
    public T getItem(int position) {
        if (adapterData == null) {
            return null;
        }
        return adapterData.get(position);
    }

    /**
     * Get the row id associated with the specified position in the list. Note that item IDs are not stable so you
     * cannot rely on the item ID being the same after {@link #notifyDataSetChanged()} or
     * {@link #updateData(OrderedRealmCollection)} has been called.
     *
     * @param position The position of the item within the adapter's data set whose row id we want.
     * @return The id of the item at the specified position.
     */
    @Override
    public long getItemId(int position) {
        // TODO: find better solution once we have unique IDs
        return position;
    }

    /**
     * Updates the data associated with the Adapter.
     *
     * Note that RealmResults and RealmLists are "live" views, so they will automatically be updated to reflect the
     * latest changes. This will also trigger {@code notifyDataSetChanged()} to be called on the adapter.
     *
     * This method is therefore only useful if you want to display data based on a new query without replacing the
     * adapter.
     *
     * @param data the new {@link OrderedRealmCollection} to display.
     */
    public void updateData(OrderedRealmCollection<T> data) {
        if (listener != null) {
            if (adapterData != null) {
                removeListener(adapterData);
            }
            if (data != null) {
                addListener(data);
            }
        }

        this.adapterData = data;
        notifyDataSetChanged();
    }
}
```

**Comment by cmelchior** [comment]
That is probably a good starting ground, but I am not sure it captures the complexity of some of the use cases we are seeing (e.g drag-n-drop sorting).

**Comment by Zhuinden** [comment]
That's true, and I think that requires stable IDs, and on top of that `notifyDatasetChanged()` also means no predictive animations on insertion/removal... 

I'm pretty sure that when we needed drag-and-drop sorting, we had detached copies (or at least a list that was not a `RealmResults<T>`), and we also had our own field called `rank`. I'm not sure this is a problem that can be handled generally, but I could be wrong.

**Comment by bkromhout** [comment]
A library I maintain, [realm-recyclerview-lite](https://github.com/bkromhout/realm-recyclerview-lite), supports drag-and-drop functionality; here's what I discovered while implementing it:
* As @Zhuinden notes, you do indeed need stable IDs. This isn't too difficult, you just need a `long` field on the model which has a unique value
* Some sort of field to track the order (such as `rank`, `position`, etc) is required, of course
* The collection that the adapter draws data from (so, `RealmResults`, for example) must *always* be up to date. Always.
* You need to support predictive animations, which means that you need a way of noticing which items have changed at the correct time

That last point is the most difficult, of course, seeing as we currently don't have fine-grained notifications. My library is a heavily changed derivative of Thorben Primke's realm-recyclerview library, and I kept the idea of using a library called java-diff-utils to help figure out what's changed when a RealmChangeListener indicates changes have occurred. 

This is done by figuring out a PK field on the model class when the adapter is created, and using that information to keep a list of PK values for all items in the `RealmResults`. This list of "IDs" is updated when the `RealmResults` are provided, and every time the change listener is called, by iterating over the items in the `RealmResults` and getting the values of their PK fields. By taking a diff of the old and the new list, the adapter is able to see where items have been removed/inserted/moved and call the correct `notify*Change()` method accordingly.

While this does indeed work in practice, there are two issues (which are related) with it, which at the current time are only solvable by having the adapter class be within the `io.realm` package and making use of internal APIs 😞 (which is why I'm hesitant to open a PR with any sort of example on this repo, since it's an official Realm offering):
* The `RealmResults` *have to* be up to date when the ID list is recreated. Prior to Realm 0.89.0 this wasn't even an issue because they were always 100% up to date. Realm 0.89.0 changed the "live-ness" of `RealmResults` such that updates are deferred to the next looper event, which causes some issues, *especially* with drag-and-drop, which is very sensitive. The only fix I could figure out was to call the `RealmResults.syncIfNeeded()` prior to rebuilding the ID list. Failing to do so resulted in weird situations where items being dragged would... well honestly, they would either just not work, or the rendering would "flip out" a bit
* While I tried my best to make use of the `Dynamic*` classes, the APIs simply just weren't what was needed to:
    * Automatically determine the PK field initially--with the change I contributed for 0.90.0, I technically could, but I would still have issues later which are too lengthy to explain right here, so I end up needing the `Table` class--...
    * And then get the values of all of the items' PK fields multiple times--i.e., when you're doing drag and drop, Android expects the underlying data to be changing with you, so while the system does batch some calls to the methods notifying you that "the dragged item has moved so you need to update your data now", worst case is that I'd need to recreate the ID list for every instance of the item being dragged past another item...in other words, it needs to be *fast*, and even the dynamic APIs wouldn't have been enough unless I used reflection (which would likely be even worse!)

Each time there's a new Realm release, I look at the changes and see if any of the APIs would make it possible to remove the internal API requirements, but currently it just doesn't look possible. The addition of fine-grained notifications (as long is the implementation is robust enough) will hopefully solve many of the issues and allow me to refactor the adapter class out of the `io.realm` package, but that remains to be seen.

TL;DR, we really need both fine-grained notifications and a way to force some realm-managed collection like `RealmResults` to sync itself no matter what in order to support something as complex as drag-and-drop on live realm collections. Currently the only alternatives are to use internal APIs (😞), or as @Zhuinden said, to have a duplicate copy of the data in a non-managed collection.

**Comment by JM-Agrimap** [comment]
Why not look at getting Thorben Primke's versions integrated as a starting point since those are the ones you reference in the "samples" section of the website?

https://github.com/thorbenprimke/realm-recyclerview/blob/master/library/src/main/java/co/moonmonkeylabs/realmrecyclerview/RealmRecyclerView.java
https://github.com/thorbenprimke/realm-recyclerview/blob/master/library/src/main/java/io/realm/RealmBasedRecyclerViewAdapter.java

**Comment by cmelchior** [comment]
Thorben has done some great work with his RecyclerView, but his library is more like a self-contained UI widget. What we want to do with the classes in this repo is making sure they can be used for building your own custom widget. For that reason his version fits better in its own repository.

**Comment by Zhuinden** [comment]
Isn't this technically https://github.com/realm/realm-android-adapters/blob/master/example/src/main/java/io/realm/examples/adapters/ui/recyclerview/MyRecyclerViewAdapter.java ?

**Comment by cmelchior** [comment]
Partly yes, but I believe this issue was also created to address the problems described here https://github.com/realm/realm-java/issues/4245#issuecomment-282432206
